### PR TITLE
Adds Drag and Drop Hover State to QuillEditor

### PR DESCRIPTION
**Pull request by martins87**
## Link to Issue
Closes: #4037 

## Description of Changes
- Adds dashed 4, 4 pattern when hovering editor
- Adds placement when image is dropped to text (blank text, beginnnig of line, middle of line, end of line, after another image)

## Test Plan
- dropping image on blank editor
- dropping image on beginning of line
- dropping image on end of line
- dropping image in the middle of line of text
- dropping image after another image

### CODE REVIEW

**Comment by jnaviask** [comment]
cc @meeshlin @jessmart1213 this is deployed on demo for acceptance

**Comment by jnaviask** [comment]
cc @zakhap as well.

**Comment by martins87** [comment]
> cc @meeshlin @jessmart1213 this is deployed on demo for acceptance

What is the demo link?

**Comment by jessmart1213** [comment]
Reviewed D&D and here is my feedback:

**DASHED LINE**
- Check: It is hard to eyeball if the implemented dashed line matches the designs in Figma. The dash pattern should be, Dash pattern: 4, 4 as seen in the Figma inspect column, of which I attached a screenshot below. 

Screen Shot 2023-06-07 at 4 07 35 PM

- I did not specify this in the original handoff, so apologies here.

**DROP PLACEMENT**
- Fix: When the cursor sits at the end or the right side of a line of text, the image should drop below that line of text.
- Fix: When the cursor sits at the beginning or the left side of a line of text, it should drop above that line of text. 

Please see loom video for reference. 

**D&D STATE SHOULD DISAPPEAR**
- Currently the drag and drop state does not disappear once a user has dragged & dropped an image and the image has loaded. As I typed, the drag and drop state stayed present. This is not the intended behavior. 

- Fix: The drag and drop state should disappear once the loading visual leaves the screen. 

Please see loom video for reference. 

**LOOM**
- [D&D Design QA Loom](https://www.loom.com/share/fc081203b632450eaebf3e1819eccc89)

Sidenote: To whomever is building this, apologies if I sound overly critical in the Loom. I appreciate the work and just a bit overwhelmed. I will be more conscientious next time.

**Comment by jnaviask** [comment]
cc @martins87 to address feedback

**Comment by martins87** [comment]
> 

Is this the exact Dash Pattern?

Site: https://kovart.github.io/dashed-border-generator/

image

**Comment by CowMuon** [comment]
What's the status here? This PR is still marked as draft and has merge conflicts @martins87

File: packages/commonwealth/client/styles/components/react_quill/react_quill_editor.scss [hunk 1]
```diff
@@ -122,6 +122,11 @@
     }
   }
 
+  .ondragover {
+    background-image: url("data:image/svg+xml,%3csvg width='100%25' height='100%25' xmlns='http://www.w3.org/2000/svg'%3e%3crect width='100%25' height='100%25' fill='none' stroke='%23338fff' stroke-width='4' stroke-dasharray='4%2c 4' stroke-dashoffset='0' stroke-linecap='butt'/%3e%3c/svg%3e");
```
Comment (masvelio) [inline packages/commonwealth/client/styles/components/react_quill/react_quill_editor.scss:126]: why using image instead of the border css?

**Comment by martins87** [comment]
> Also the spinner is off the editor when the image is being updated. Should be in the middle.

I'm on it.

**Comment by masvelio** [comment]
> For some reason I am not able to upload an image. Do you have that in you local env as well? Worth adding some notification error actually that `Upload failed`. (If not in scope, good FF ticket)
> 
>  Kapture.2023-07-20.at.15.49.21.mp4 
> Also please check if [this still happens ](https://github.com/hicommonwealth/commonwealth/pull/4542#pullrequestreview-1539285646)

Ok, upload works on my side - I had env vars messed up. 
But uploading more than one picture still does not work as I described here https://github.com/hicommonwealth/commonwealth/pull/4542#pullrequestreview-1539285646

**Comment by masvelio** [comment]
@martins87 still not ready to be merged
1. conflicts
2. loading spinner is not centered and is underneath the editor
3. bug => upload first image, upload second image, first image is gone. This works fine on master

https://github.com/hicommonwealth/commonwealth/assets/14819225/6e69ceb6-4390-4001-8eb9-e21f348adc98

cc @MuonShot @jnaviask

**Comment by martins87** [comment]
> @martins87 still not ready to be merged
> 
> 1. conflicts
> 2. loading spinner is not centered and is underneath the editor
> 3. bug => upload first image, upload second image, first image is gone. This works fine on master
> 
>  Kapture.2023-08-01.at.11.59.29.mp4 
> cc @MuonShot @jnaviask

1. Conflicts solved.
2. Fixed loading spinner: on center of Quill Editor
3. Upload of mutiples images ok

**Comment by martins87** [comment]
Some tests:

1. Blank editor
2. End of line
3. Middle of line
4. After an image
5. Multiple images

https://github.com/hicommonwealth/commonwealth/assets/30223098/05f7fab3-a9af-4bf8-a085-1c436f3d80b3

**Comment by martins87** [comment]
Note on the dash pattern: the current implementation follows the one Jessica commented (Dash Pattern: 4, 4)

If it does not look ideal, the below dashed blue border can be obtained with `border: 2px dashed $primary-500;`

image

**Comment by jnaviask** [comment]
Let's get this on demo for design review + final approval. cc @Miaplacidus

File: packages/commonwealth/client/scripts/views/components/react_quill_editor/use_image_drop_and_paste.tsx [hunk 2]
```diff
@@ -72,11 +69,32 @@ export const useImageDropAndPaste = ({
...
         } else {
-          editor.insertEmbed(selectedIndex, 'image', uploadedFileUrl);
+          if (!blankEditor) {
+            editor.insertText(selectedIndex, lineBreak);
+          }
+          if (selectedIndex === 0) {
+            editor.insertEmbed(selectedIndex, 'image', uploadedFileUrl);
+          } else {
+            editor.insertEmbed(selectedIndex + 1, 'image', uploadedFileUrl);
+          }
```
Comment (masvelio) [inline packages/commonwealth/client/scripts/views/components/react_quill_editor/use_image_drop_and_paste.tsx:97]: please add comments in the code what cases these if statements cover

**Comment by masvelio** [comment]
codewise looks good, let's get the sign off from the design team

File: packages/commonwealth/client/scripts/views/components/react_quill_editor/use_image_drop_and_paste.tsx (hunk 2 above)
Comment (jnaviask) [reply packages/commonwealth/client/scripts/views/components/react_quill_editor/use_image_drop_and_paste.tsx:97]: cc @martins87

**Comment by masvelio** [comment]
Todo:
- [ ] resolve conflicts
- [ ] CORS problem on demo environment on image upload
- [ ] design sign off

File: packages/commonwealth/client/scripts/views/components/react_quill_editor/use_image_drop_and_paste.tsx (hunk 2 above)
Comment (martins87) [reply packages/commonwealth/client/scripts/views/components/react_quill_editor/use_image_drop_and_paste.tsx:99]: [Addressed](https://github.com/hicommonwealth/commonwealth/pull/4109/commits/a59211db3ed988a1cbfc461f9a81bda89a9c3806).

**Comment by martins87** [comment]
> Todo:
> 
> * [x]  resolve conflicts
> * [x]  CORS problem on demo environment on image upload
> * [ ]  design sign off

@jessmart1213, I've added the edits to demo environment. Please check them on [/components](https://commonwealth-demo.herokuapp.com/components) whenever you can.

**Comment by masvelio** [comment]
@jessmart1213 could you take a look on https://commonwealth-demo.herokuapp.com/ if drag&drop is satisfying now?

**Comment by masvelio** [comment]
what's the status here?

**Comment by jessmart1213** [comment]
The drag and drop hover state has Deign sign-off! Thank you @martins87!

**Screen-recording of QA**
https://github.com/hicommonwealth/commonwealth/assets/133259695/0e7c559c-8858-45c0-bf7b-3fb7adbba6fd

**Review by masvelio — APPROVED** [review]

**Comment by martins87** [comment]
> The drag and drop hover state has Deign sign-off! Thank you @martins87!
> 
> **Screen-recording of QA** https://github.com/hicommonwealth/commonwealth/assets/133259695/0e7c559c-8858-45c0-bf7b-3fb7adbba6fd

@jnaviask

**Review by jnaviask — APPROVED** [review]

**Pull request merged**